src/Symfony/Component/Utf8/CodePoints.php

```diff
@@ -0,0 +1,390 @@
+            $chunks[$i] = $chunk = clone $this;
+            $chunk->string = $string;
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtolower($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtoupper($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function substr(int $start = 0, int $length = null): self
+    {
+        $result = clone $this;
+        $result->string = mb_substr($this->string, $start, $length, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
+
+        $result = clone $this;
+        $result->string = preg_replace("/^[$charsList]+|[$charsList]+\$/u", '', $this->string);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimLeft(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
+
+        $result = clone $this;
+        $result->string = preg_replace("/^[$charsList]+/u", '', $this->string);
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimRight(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
+
+        $result = clone $this;
+        $result->string = preg_replace("/[$charsList]+\$/u", '', $this->string);
+
+        return $result;
+    }
+
+    public function getIterator(int $maxChunkLength = 1): \Generator
+    {
+        if ($maxChunkLength < 1) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return;
+        }
+
+        if ($maxChunkLength > 65535) {
+            $maxChunkLength = 65535;
+        }
+
+        foreach (preg_split('/(.{'.$maxChunkLength.'})/u', $this->string, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY) as $key => $char) {
+            $clone = clone $this;
+            $clone->string = $char;
+            yield $key => $clone;
+        }
+    }
+
+    public function append(string $suffix): self
+    {
+        $result = clone $this;
+        $result->string.= self::fromString($suffix)->string;
+
+        return $result;
+    }
+
+    public function prepend(string $prefix): self
+    {
+        $result = self::fromString($prefix);
+        $result->string.= $this->string;
+
+        return $result;
+    }
+
+    public function toBytes(): Bytes
```

Comment:
add this (and the two others) to the interface? toCodePoints would just return $this here